linalg_ops/attention/generate_e2e_attention_tests.py

```diff
@@ -0,0 +1,503 @@
+@dataclasses.dataclass
+class MLIRFunction:
+    name: str
+    signature: str
+    import_declaration: str
+    definition: str
+
+
+# Generates a test function in the generated MLIR code.
+# The generated function will take the same arguments as iree_linalg_ext.attention variants
+# and will just call iree_linalg_ext.attention variants with them, returning its result.
+def generate_function(
+    query_type: QueryElemTypeId,
+    key_type: KeyElemTypeId,
+    value_type: ValueElemTypeId,
+    shape_scale: TestShapeAndScale,
+):
+    shapes_scale = generate_shapes_and_scale(shape_scale)
+    func_name = generate_function_name(
+        query_type,
+        key_type,
+        value_type,
+        shapes_scale,
+    )
+
+    query_shape, key_shape, value_shape, result_shape = get_tensor_shapes(shapes_scale)
+    query_tensor_type = (
+        f"tensor<{query_shape[0]}x{query_shape[1]}x{query_shape[2]}x{query_type.value}>"
+    )
+    key_tensor_type = (
+        f"tensor<{key_shape[0]}x{key_shape[1]}x{key_shape[2]}x{key_type.value}>"
+    )
+    value_tensor_type = (
+        f"tensor<{value_shape[0]}x{value_shape[1]}x{value_shape[2]}x{value_type.value}>"
+    )
+    result_tensor_type = f"tensor<{result_shape[0]}x{result_shape[1]}x{result_shape[2]}x{value_type.value}>"
+    F32 = "f32"
+    F16 = "f16"
+    op_name = "iree_linalg_ext.attention"
+
+    # Compilation info is optional; prints empty string by default.
+    func_definition = ""
+
+    signature = f"({query_tensor_type}, {key_tensor_type}, {value_tensor_type}, {result_tensor_type}) -> {result_tensor_type}"
+    import_declaration = f"func.func private @module.{func_name}(%query: !hal.buffer_view, %key: !hal.buffer_view, %value: !hal.buffer_view, %scale: {F32}) -> !hal.buffer_view"
+    func_definition = func_definition + (
+        f"func.func @{func_name}(%query: {query_tensor_type}, %key: {key_tensor_type}, %value: {value_tensor_type}, %scale: {F32}) -> {result_tensor_type} {{\n"
+        f"  %result0 = tensor.empty(): {result_tensor_type}\n"
+        f"  %scale_f16 = arith.truncf %scale : {F32} to {F16} \n"
+        f"  %result1 = {op_name} {{\n"
+        f"      indexing_maps = [affine_map<(batch, m, n, k1, k2) -> (batch, m, k1)>,\n"
+        f"                       affine_map<(batch, m, n, k1, k2) -> (batch, k2, k1)>,\n"
+        f"                       affine_map<(batch, m, n, k1, k2) -> (batch, k2, n)>,\n"
+        f"                       affine_map<(batch, m, n, k1, k2) -> ()>,\n"
+        f"                       affine_map<(batch, m, n, k1, k2) -> (batch, m, n)>]\n}}"
+        f"      ins(%query, %key, %value, %scale_f16: {query_tensor_type}, {key_tensor_type}, {value_tensor_type}, {F16})\n"
+        f"      outs(%result0: {result_tensor_type}) {{\n"
+        f"   ^bb0(%score: f32): \n"
+        f"   iree_linalg_ext.yield %score : f32\n"
+        f" }} -> {result_tensor_type}\n"
+        f" return %result1: {result_tensor_type}\n"
+        f"}}\n"
```

Comment:
nit: Is this `\n` needed? Looks like the generated files have an extra

![image](https://github.com/user-attachments/assets/c69a0d58-4a22-4e92-a355-7be7699e1af2)